daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala

```diff
@@ -93,88 +95,71 @@ object SExpr {
           SPAP(PBuiltin(b), new util.ArrayList(), b.arity)
       }
     }
-    def execute(machine: Machine): Unit = {
-      /* special case for nullary record constructors */
-      machine.returnValue = b match {
-        case SBRecCon(id, fields) if b.arity == 0 =>
-          SRecord(id, fields, new util.ArrayList())
-        case _ =>
-          SPAP(PBuiltin(b), new util.ArrayList(), b.arity)
-      }
-    }
   }
 
   /** A pre-computed value, usually primitive literal, e.g. integer, text, boolean etc. */
   final case class SEValue(v: SValue) extends SExprAtomic {
-    def evaluate(machine: Machine): SValue = {
+    def lookupValue(machine: Machine): SValue = {
       v
     }
-    def execute(machine: Machine): Unit = {
-      machine.returnValue = v
-    }
   }
 
   object SEValue extends SValueContainer[SEValue]
 
-  /** Function application:
-    General case: 'fun' and 'args' are any kind of expression */
+  /** Function application: General case: 'fun' and 'args' are any kind of expression */
   final case class SEAppGeneral(fun: SExpr, args: Array[SExpr]) extends SExpr with SomeArrayEquals {
     def execute(machine: Machine): Unit = {
-      machine.pushKont(KArg(args, machine.frame, machine.actuals, machine.env.size))
-      machine.ctrl = fun
+      // We never encounter a general application node when executing on the speedy machine
+      // This is because all speedy expression must have been converted to ANF form
+      // The wrapper type `AExpr` helps ensure this restrction is adhered to.
+      // Better still would be to properly separate the `SExpr` and `AExpr` types.
```

Comment:
Is this still true even when only partially transforming to ANF?

Reply:
YES! We are still transforming to satisfy the ANF invariant. But just less optimally. Preserving this invariant is the reason to take the approach of transforming unknown multi-apps into a cascade of single-apps. If we don't do this, we have to reinstate code in the speedy engine for handling non-ANF applications.